internal/report/store_memory.go

```diff
@@ -0,0 +1,198 @@
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	report, found := s.prCache[namespace]
+
+	if !found {
+		return PolicyReport{}, constants.ErrResourceNotFound
+	}
+
+	log.Debug().Dict("dict", zerolog.Dict().
+		Str("report name", report.GetName()).
+		Str("report ns", report.GetNamespace()).
+		Str("report resourceVersion", report.GetResourceVersion())).
+		Msg("PolicyReport found")
+	return report, nil
+}
+
+func (s *MemoryPolicyReportStore) GetClusterPolicyReport(name string) (ClusterPolicyReport, error) {
+	if !strings.HasPrefix(name, PrefixNameClusterPolicyReport) {
+		name = getClusterReportName(name)
+	}
+
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	report, found := s.cprCache[name]
+
+	if !found {
+		return ClusterPolicyReport{}, constants.ErrResourceNotFound
+	}
+
+	return report, nil
+}
+
+func (s *MemoryPolicyReportStore) UpdatePolicyReport(report *PolicyReport) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.prCache[report.GetNamespace()] = *report
+
+	summary, _ := report.GetSummaryJSON()
+	log.Info().
+		Dict("dict", zerolog.Dict().
+			Str("report name", report.GetName()).
+			Str("report ns", report.GetNamespace()).
+			Str("report resourceVersion", report.GetResourceVersion()).
+			Str("summary", summary),
+		).Msg("updated PolicyReport")
+	return nil
+}
+
+func (s *MemoryPolicyReportStore) UpdateClusterPolicyReport(report *ClusterPolicyReport) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.cprCache[report.GetName()] = *report
+
+	summary, _ := report.GetSummaryJSON()
+	log.Info().
+		Dict("dict", zerolog.Dict().
+			Str("report name", report.GetName()).
+			Str("report ns", report.GetNamespace()).
+			Str("summary", summary),
+		).Msg("updated ClusterPolicyReport")
+	return nil
+}
+
+func (s *MemoryPolicyReportStore) RemovePolicyReport(namespace string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	delete(s.prCache, namespace)
+
+	return nil
+}
+
+func (s *MemoryPolicyReportStore) RemoveAllNamespacedPolicyReports() error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.prCache = make(map[string]PolicyReport)
+
+	return nil
+}
+
+func (s *MemoryPolicyReportStore) SavePolicyReport(report *PolicyReport) error {
+	// Check for existing Policy Report
+	_, getErr := s.GetPolicyReport(report.GetNamespace())
+	if getErr != nil {
+		// Create new Policy Report if not found
+		if errors.Is(getErr, constants.ErrResourceNotFound) {
+			// Update will create a new one if it doesn't exist
+			return s.UpdatePolicyReport(report)
+		}
+		return getErr
+	}
+	// Update existing Policy Report
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// get the latest report version to be updated
+		latestReport, err := s.GetPolicyReport(report.GetNamespace())
+		if err != nil {
+			return err
+		}
+		latestReport.Summary = report.Summary
+		latestReport.Results = report.Results
+		return s.UpdatePolicyReport(&latestReport)
+	})
```

Comment:
I think you can remove the `RetryOnConflict`. This is use in K8s report because of the conflict it can happen when updating the report in the backend. Here, I don't expect that to happen so often. Therefore, checking the `GetPolicyReport` error once is enough.

Reply:
that makes sense, thank you for noticing that! changed for `SavePolicyReport` and [SaveClusterPolicyReport](https://github.com/kubewarden/audit-scanner/pull/150#discussion_r1417270463)